File: src/core/task.cpp
// SPDX-FileCopyrightText: 2025 Sergio Martins
// SPDX-License-Identifier: MIT

#include "task.h"
#include "logger.h"
#include "tag.h"
#include "Clock.h"
#include "date_utils.h"

#include <algorithm>
#include <iomanip>
#include <sstream>

namespace pointless::core {

Task::Task() = default;

Task::Task(std::string uuid_, std::chrono::system_clock::time_point creationTimestamp_, std::string title_)
    : uuid(std::move(uuid_))
    , title(std::move(title_))
    , creationTimestamp(creationTimestamp_)
{
}

bool Task::containsTag(std::string_view tagName) const
{
    return std::ranges::any_of(tags, [tagName](const auto &tag) {
        return tag == tagName;
    });
}

bool Task::isSoon() const
{
    if (isCurrent())
        return false;

    return containsTag(BUILTIN_TAG_SOON) || isDueIn(std::chrono::days(15));
}

bool Task::isLater() const
{
    return !isSoon() && !isCurrent();
}

bool Task::isCurrent() const
{
    return containsTag(BUILTIN_TAG_CURRENT) || isDueThisWeek();
}

bool Task::isEvening() const
{
    return containsTag(BUILTIN_TAG_EVENING);
}

std::string Task::tagName() const
{
    for (const auto &tag : tags) {
        if (!tagIsBuiltin(tag)) {
            return tag;
        }
    }
    return {};
}

bool Task::isDueIn(std::chrono::days days) const
{
    if (!dueDate) {
        return false;
    }
    const auto now = Clock::now();
    const auto due = *dueDate;
    return due <= now + days && due >= now;
}

bool Task::isDueTomorrow() const
{
    if (!dueDate) {
        return false;
    }
    return DateUtils::isTomorrow(*dueDate);
}

bool Task::isDueThisWeek() const
{
    if (!dueDate) {
        return false;
    }
    return DateUtils::isThisWeek(*dueDate);
}

void Task::setTags(const std::vector<std::string> &newTags)
{
    tags = {};
    for (const auto &tag : newTags) {
        addTag(tag);
    }
}

bool Task::addTag(std::string_view tag)
{
    if (!tag.empty() && !containsTag(tag)) {
        tags.emplace_back(tag);
        return true;
    }
    return false;
}

void Task::removeBuiltinTags()
{
    std::erase_if(tags, [](const std::string &tag) {
        return tagIsBuiltin(tag);
    });
}

void Task::dumpDebug() const
{
    std::string json;
    if (glz::write_json(*this, json)) {
        P_LOG_ERROR("Failed to serialize task to JSON");
    } else {
        auto pretty = glz::prettify_json(json);
        P_LOG_DEBUG("Task JSON: {}", pretty);
    }

    auto timeToString = [](std::chrono::system_clock::time_point tp) {
        auto t = std::chrono::system_clock::to_time_t(tp);
        std::stringstream ss;
        ss << std::put_time(std::localtime(&t), "%Y-%m-%d %H:%M:%S");
        return ss.str();
    };

    P_LOG_DEBUG("Creation: {}", timeToString(creationTimestamp));
    if (modificationTimestamp)
        P_LOG_DEBUG("Modification: {}", timeToString(*modificationTimestamp));
    if (lastPomodoroDate)
        P_LOG_DEBUG("Last Pomodoro: {}", timeToString(*lastPomodoroDate));
    if (dueDate)
        P_LOG_DEBUG("Due: {}", timeToString(*dueDate));
    if (completionDate)
        P_LOG_DEBUG("Completion: {}", timeToString(*completionDate));
}


void Task::mergeConflict(const Task &other)
{
    // Completion: Undone > Done
    isDone = isDone && other.isDone;

    // Importance: Important > Not Important
    isImportant = isImportant || other.isImportant;

    auto myTime = modificationTimestamp.value_or(std::chrono::system_clock::time_point::min());
    auto otherTime = other.modificationTimestamp.value_or(std::chrono::system_clock::time_point::min());

    const bool otherIsMoreRecent = otherTime > myTime;

    // Due Date
    if (dueDate.has_value() && other.dueDate.has_value()) {
        if (otherIsMoreRecent) {
            dueDate = other.dueDate;
        }
    } else if (other.dueDate.has_value()) {
        dueDate = other.dueDate;
    }

    // Title
    if (otherIsMoreRecent && other.title != title) {
        title = other.title;
    }

    // Tags: Union
    std::vector<std::string> newTags = tags;
    for (const auto &tag : other.tags) {
        newTags.push_back(tag);
    }
    setTags(newTags);

    // Special Rule: Current wins over Soon
    bool hasCurrent = containsTag(BUILTIN_TAG_CURRENT);
    bool hasSoon = containsTag(BUILTIN_TAG_SOON);

    if (hasCurrent && hasSoon) {
        // Remove Soon
        std::erase(tags, std::string(BUILTIN_TAG_SOON));
    }
}
} // namespace pointless::core

File: electron/src/core/task.js
// SPDX-FileCopyrightText: 2025 Sergio Martins
// SPDX-License-Identifier: MIT

'use strict';

const { tagIsBuiltin, BUILTIN_TAG_SOON, BUILTIN_TAG_CURRENT, BUILTIN_TAG_EVENING } = require('./tag');

function _tsFromMillis(value) {
    if (value === null || value === undefined) return null;
    return new Date(value);
}

function _trimTime(date) {
    const d = new Date(date);
    d.setHours(0, 0, 0, 0);
    return d;
}

function _today() {
    return _trimTime(new Date());
}

function _thisWeeksMonday(date) {
    const d = _trimTime(date);
    const day = d.getDay(); // 0=Sun
    const daysBack = day === 0 ? 6 : day - 1;
    d.setDate(d.getDate() - daysBack);
    return d;
}

function _nextMonday(monday) {
    const d = new Date(monday);
    d.setDate(d.getDate() + 7);
    return d;
}

function _isThisWeek(date) {
    const monday = _thisWeeksMonday(new Date());
    return date >= monday && date < _nextMonday(monday);
}

function _isTomorrow(date) {
    const trimmed = _trimTime(date);
    const tomorrow = _today();
    tomorrow.setDate(tomorrow.getDate() + 1);
    return trimmed.getTime() === tomorrow.getTime();
}

class Task {
    constructor({
        revision = -1,
        needsSyncToServer = false,
        uuid = '',
        parentUuid = null,
        title = '',
        isDone = false,
        isGoal = null,
        isYearly = null,
        isImportant = false,
        hideOnWeekends = null,
        timesPerWeek = 1,
        lastCompletions = [],
        sectionName = '',
        tags = [],
        creationTimestamp = null,
        modificationTimestamp = null,
        lastPomodoroDate = null,
        dueDate = null,
        completionDate = null,
        uuidInDeviceCalendar = null,
        deviceCalendarUuid = null,
        deviceCalendarName = null,
        description = null,
    } = {}) {
        this.revision = revision;
        this.needsSyncToServer = needsSyncToServer;
        this.uuid = uuid;
        this.parentUuid = parentUuid;
        this.title = title;
        this.isDone = isDone;
        this.isGoal = isGoal;
        this.isYearly = isYearly;
        this.isImportant = isImportant;
        this.hideOnWeekends = hideOnWeekends;
        this.timesPerWeek = timesPerWeek;
        this.lastCompletions = lastCompletions;
        this.sectionName = sectionName;
        this.tags = tags;
        this.creationTimestamp = creationTimestamp;
        this.modificationTimestamp = modificationTimestamp;
        this.lastPomodoroDate = lastPomodoroDate;
        this.dueDate = dueDate;
        this.completionDate = completionDate;
        this.uuidInDeviceCalendar = uuidInDeviceCalendar;
        this.deviceCalendarUuid = deviceCalendarUuid;
        this.deviceCalendarName = deviceCalendarName;
        this.description = description;
    }

    containsTag(tagName) {
        return this.tags.includes(tagName);
    }

    isDueIn(days) {
        if (!this.dueDate) return false;
        const now = Date.now();
        const due = this.dueDate.getTime();
        return due <= now + days * 86400000 && due >= now;
    }

    isDueTomorrow() {
        if (!this.dueDate) return false;
        return _isTomorrow(this.dueDate);
    }

    isDueThisWeek() {
        if (!this.dueDate) return false;
        return _isThisWeek(this.dueDate);
    }

    isOverdue() {
        if (!this.dueDate || this.isDone) return false;
        return this.dueDate.getTime() < Date.now();
    }

    isCurrent() {
        return this.containsTag(BUILTIN_TAG_CURRENT) || this.isDueThisWeek() || this.isOverdue();
    }

    isSoon() {
        if (this.isCurrent()) return false;
        return this.containsTag(BUILTIN_TAG_SOON) || this.isDueIn(15);
    }

    isLater() {
        return !this.isSoon() && !this.isCurrent();
    }

    isEvening() {
        return this.containsTag(BUILTIN_TAG_EVENING);
    }

    shouldBeCleanedUp() {
        if (this.isYearly === true) return false;
        const twoWeeksAgo = Date.now() - 14 * 86400000;
        if (!this.isDone) return false;
        if (this.modificationTimestamp) {
            return this.modificationTimestamp.getTime() < twoWeeksAgo;
        }
        return true;
    }

    tagName() {
        for (const tag of this.tags) {
            if (!tagIsBuiltin(tag)) return tag;
        }
        return '';
    }

    addTag(tag) {
        if (tag && !this.containsTag(tag)) {
            this.tags.push(tag);
            return true;
        }
        return false;
    }

    removeBuiltinTags() {
        this.tags = this.tags.filter(t => !tagIsBuiltin(t));
    }

    static fromJSON(obj) {
        return new Task({
            revision: obj.revision ?? -1,
            needsSyncToServer: obj.needsSyncToServer ?? false,
            uuid: obj.uuid ?? '',
            parentUuid: obj.parentUuid ?? null,
            title: obj.title ?? '',
            isDone: obj.isDone ?? false,
            isGoal: obj.isGoal ?? null,
            isYearly: obj.isYearly ?? null,
            isImportant: obj.isImportant ?? false,
            hideOnWeekends: obj.hideOnWeekends ?? null,
            timesPerWeek: obj.timesPerWeek ?? 1,
            lastCompletions: obj.lastCompletions ?? [],
            sectionName: obj.sectionName ?? '',
            tags: obj.tags ?? [],
            creationTimestamp: _tsFromMillis(obj.creationTimestamp),
            modificationTimestamp: _tsFromMillis(obj.modificationTimestamp),
            lastPomodoroDate: _tsFromMillis(obj.lastPomodoroDate),
            dueDate: _tsFromMillis(obj.dueDate),
            completionDate: _tsFromMillis(obj.completionDate),
            uuidInDeviceCalendar: obj.uuidInDeviceCalendar ?? null,
            deviceCalendarUuid: obj.deviceCalendarUuid ?? null,
            deviceCalendarName: obj.deviceCalendarName ?? null,
            description: obj.description ?? null,
        });
    }
}

module.exports = { Task };
